source/common/config/grpc_mux_impl.cc

```diff
@@ -47,10 +51,32 @@ void GrpcMuxImpl::establishNewStream() {
 
   control_plane_stats_.connected_state_.set(1);
   for (const auto type_url : subscriptions_) {
-    sendDiscoveryRequest(type_url);
+    queueDiscoveryRequest(type_url);
+  }
+}
+
+void GrpcMuxImpl::drainRequests() {
+  ENVOY_LOG(trace, "draining discovery requests {}", request_queue_.size());
+  while (!request_queue_.empty()) {
+    // Process the request if it is under rate limit.
+    if (limit_request_->consume()) {
+      sendDiscoveryRequest(request_queue_.front());
+      request_queue_.pop();
+    } else {
+      ENVOY_LOG(warn, "Too many sendDiscoveryRequest calls");
+      // Enable the drain request timer.
+      drain_request_timer_->enableTimer(
```

Comment:
Is there a test that verifies that we end up sleeping for this specific period. I can't spot it in the diff.

Reply:
That test is not there. I thought TokenBucketImpl test validate the period. But I will add one to be sure here as well.